### ISSUE TITLE
Imagick plugin: Option to filter transformations based on source file information

### ISSUE THREAD

**Issue by paulrobertlloyd** [lumeland/lume#279]
I have a number of images, some of which I want to perform additional or different transformations on. Right now, this isn’t possible without saving files to a different directories with different directory data.

For example, for a given blog post, say I have a folder of associated images. To enable responsive images, I may want to generate different formats and/or sizes for each of these images. However, one of these images is also the ‘key’ image. For this image, I also want to:

* Generate a cropped 1200×630 JPEG, with suffix `-summary`, to use for the post’s Open Graph image
* Generate cropped 600×400 JPEG and WEBP images to use on the articles index page

To achieve this, I either need to move/copy this ‘key’ image to a different folder (possible, but inflexible) or perform transformations on *all* the images in the same directory – an expensive file operation that also increases the number and size of files needing to be managed during deployment.

One possible way to achieve the above would be to provide a matching mechanism in the `imagick` data, under a new key called `matches`[^1]:

```yaml
imagick:
  - suffix: -summary
    matches: (.*)-main.jpg
    resizeCrop: [1200, 630]
    format: jpg
  - suffix: -big
    resize: [2000, 0]
    format: [jpg, webp]
  - resize: [1000, 0]
    format: [jpg, webp]
```

I’m not sure what the right syntax for determine matches would be, and it’s hard to envisage it would be able to cover all possible use cases. Perhaps a better idea would be to allow `matches` to accept a function (meaning you’d need to use a JavaScript data file):

```js
export default {
  imagick: [{
    suffix: "-summary",
    matches: (src) => src.path.includes("main.jpg"),
    resizeCrop: [1200, 630],
    format: "jpg",
  }, {
    suffix: "-big",
    resize: [2000, 0],
    format: ["jpg", "webp"],
  }, {
    resize: [1000, 0],
    format: ["jpg", "webp"],
  }],
};
```

Alternatively, maybe filters can query file information about the image:

```js
site.use(imagick({
  functions: {
    summaryImage({ img, width }, file) {
      if (file.src.path.includes("item.jpg")) {
        img.resize(width, 0);
        img.crop(width, height, 5);
      }
    },
  },
}));
```

[^1]: `matches` is the first name that came to mind. I thought about `filter`, but this might get confused with graphic image filters.

**Comment by oscarotero** [comment]
Ok, I've implemented `matches` which accepts a `RegEx` or a `string` (that is used to create a RegEx instance).
Now, you can do:

```yml
imagick:
  - suffix: -summary
    matches: -main.jpg
    resizeCrop: [1200, 630]
    format: jpg
```

It's included in the coming Lume 1.12.0.

**Comment by paulrobertlloyd** [comment]
Amazing!! Thank-you so much for the quick updates. Excited to try out the next version 😄
